File: src/NerdBank.GitVersioning/CloudBuildServices/AppVeyor.cs
// Copyright (c) .NET Foundation and Contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System.ComponentModel;
using System.Diagnostics;

namespace Nerdbank.GitVersioning.CloudBuildServices;

/// <summary>
/// Cloud build handling for AppVeyor.
/// </summary>
/// <remarks>
/// The AppVeyor-specific properties referenced here are <see href="http://www.appveyor.com/docs/environment-variables">documented here</see>.
/// </remarks>
internal class AppVeyor : ICloudBuild
{
    /// <inheritdoc/>
    /// <remarks>
    /// AppVeyor's branch variable is the target branch of a PR, which is *NOT* to be misinterpreted
    /// as building the target branch itself. So only set the branch built property if it's not a PR.
    /// </remarks>
    public string BuildingBranch => !this.IsPullRequest && !string.IsNullOrEmpty(Environment.GetEnvironmentVariable("APPVEYOR_REPO_BRANCH"))
        ? $"refs/heads/{Environment.GetEnvironmentVariable("APPVEYOR_REPO_BRANCH")}"
        : null;

    public string BuildingRef => null;

    /// <inheritdoc/>
    public string BuildingTag => string.Equals("true", Environment.GetEnvironmentVariable("APPVEYOR_REPO_TAG"), StringComparison.OrdinalIgnoreCase)
        ? $"refs/tags/{Environment.GetEnvironmentVariable("APPVEYOR_REPO_TAG_NAME")}"
        : null;

    /// <inheritdoc/>
    public string GitCommitId => null;

    /// <inheritdoc/>
    public bool IsApplicable => !string.IsNullOrEmpty(Environment.GetEnvironmentVariable("APPVEYOR"));

    /// <inheritdoc/>
    public bool IsPullRequest => !string.IsNullOrEmpty(Environment.GetEnvironmentVariable("APPVEYOR_PULL_REQUEST_NUMBER"));

    /// <inheritdoc/>
    public IReadOnlyDictionary<string, string> SetCloudBuildNumber(string buildNumber, TextWriter stdout, TextWriter stderr)
    {
        // We ignore exit code so as to not fail the build when the cloud build number is not unique.
        RunAppveyor($"UpdateBuild -Version \"{buildNumber}\"", stdout, stderr);
        return new Dictionary<string, string>();
    }

    /// <inheritdoc/>
    public IReadOnlyDictionary<string, string> SetCloudBuildVariable(string name, string value, TextWriter stdout, TextWriter stderr)
    {
        RunAppveyor($"SetVariable -Name {name} -Value \"{value}\"", stdout, stderr);
        return new Dictionary<string, string>();
    }

    private static void RunAppveyor(string args, TextWriter stdout, TextWriter stderr)
    {
        try
        {
            // Skip this if this build is running in our own unit tests, since that can
            // mess with AppVeyor's actual build information.
            if (string.IsNullOrEmpty(Environment.GetEnvironmentVariable("_NBGV_UnitTest")))
            {
                Process.Start("appveyor", args)
                    .WaitForExit();
            }
        }
        catch (Win32Exception ex) when ((uint)ex.HResult == 0x80004005)
        {
            (stderr ?? Console.Error).WriteLine("Could not find appveyor tool to set cloud build variable.");
        }
    }
}

File: src/nerdbank-gitversioning-rs/src/cloud_build_services/appveyor.rs
// Copyright (c) .NET Foundation and Contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

//! [AppVeyor](https://www.appveyor.com/docs/environment-variables/) support.

use std::io::{self, ErrorKind, Write};
use std::process::Command;

use crate::cloud::{CloudBuild, EnvironmentVariables, empty_variables, env, is_nonempty};

/// AppVeyor cloud-build integration.
pub struct AppVeyor;

impl AppVeyor {
    fn run(arguments: &[&str], stderr: &mut dyn Write) -> io::Result<()> {
        if is_nonempty(env("_NBGV_UnitTest").as_deref()) {
            return Ok(());
        }

        match Command::new("appveyor").args(arguments).status() {
            Ok(_) => Ok(()),
            Err(error) if error.kind() == ErrorKind::NotFound => {
                writeln!(
                    stderr,
                    "Could not find appveyor tool to set cloud build variable."
                )
            }
            Err(error) => Err(error),
        }
    }
}

impl CloudBuild for AppVeyor {
    fn name(&self) -> &'static str {
        "AppVeyor"
    }

    fn is_applicable(&self) -> bool {
        is_nonempty(env("APPVEYOR").as_deref())
    }

    fn is_pull_request(&self) -> bool {
        is_nonempty(env("APPVEYOR_PULL_REQUEST_NUMBER").as_deref())
    }

    fn building_branch(&self) -> Option<String> {
        (!self.is_pull_request())
            .then(|| env("APPVEYOR_REPO_BRANCH"))
            .flatten()
            .filter(|branch| !branch.is_empty())
            .map(|branch| format!("refs/heads/{branch}"))
    }

    fn building_tag(&self) -> Option<String> {
        env("APPVEYOR_REPO_TAG")
            .is_some_and(|value| value.eq_ignore_ascii_case("true"))
            .then(|| {
                format!(
                    "refs/tags/{}",
                    env("APPVEYOR_REPO_TAG_NAME").unwrap_or_default()
                )
            })
    }

    fn git_commit_id(&self) -> Option<String> {
        None
    }

    fn set_cloud_build_number(
        &self,
        build_number: &str,
        _stdout: &mut dyn Write,
        stderr: &mut dyn Write,
    ) -> io::Result<EnvironmentVariables> {
        Self::run(&["UpdateBuild", "-Version", build_number], stderr)?;
        Ok(empty_variables())
    }

    fn set_cloud_build_variable(
        &self,
        name: &str,
        value: &str,
        _stdout: &mut dyn Write,
        stderr: &mut dyn Write,
    ) -> io::Result<EnvironmentVariables> {
        Self::run(&["SetVariable", "-Name", name, "-Value", value], stderr)?;
        Ok(empty_variables())
    }
}
